File: mirea/cources/beginner/connected-components-3/e.py
# структура для системы непересекающихся множеств
class DSU:
    
    def __init__(self, n):
        self.sz = [1] * n;
        self.par = [int(x) for x in range(n)];
        self.numComponents = n;
    
    def getPar(self, u):
        if u == self.par[u]:
            return u;
        p = self.getPar(self.par[u]);
        self.par[u] = p;
        return p;
    
    def connect(self, u, v):
        u, v = self.getPar(u), self.getPar(v);
        if u != v:
            if self.sz[u] < self.sz[v]:
                u, v = v, u; # swap
            self.sz[u] += self.sz[v];
            self.par[v] = u;
            self.numComponents -= 1;

# решение задачи
t = int(input())
for id in range(t):
    n = int(input())
    a = [int(x)-1 for x in input().split()]
    # максимальный ответ: соединим только с теми соседями, которых
    # запомнили. Тогда количество компонент и будет ответом
    dsu = DSU(n);
    for i in range(n):
        dsu.connect(i, a[i]);
    # теперь считаем минимальный ответ
    # есть компоненты-деревья и компоненты-циклы
    # компоненты-циклы нельзя ни с кем соединять
    # компоненты-деревья можем соединить друг с другом
    # получается, минимальный ответ:
    # кол-во компонент-циклов + (есть ли компоненты-деревья)
    # для каждой компоненты посчитаем число уникальных рёбер
    # вычитая количество петель
    uniqEdges = [0] * n;
    isParent = [False] * n;
    for i in range(n):
        pi = dsu.getPar(i);
        uniqEdges[pi] = 2 * dsu.sz[pi];
        isParent[pi] = True;
    # вычитаем петли
    for u in range(n):
        # есть ребро u->a[u]. Проверим, есть ли ребро a[u]->u
        if (a[a[u]] == u): # петля
            uniqEdges[dsu.getPar(u)] -= 1; # вычитаем один раз
                                    # второй раз вычтется симметрично
    # теперь циклы это компонента в которой sz[i] ребёр без петель
    # остальное - деревья
    nCycles, nTrees = 0, 0;
    for i in range(n):
        if isParent[i]:
            if uniqEdges[i] == 2 * dsu.sz[i]:
                nCycles += 1;
            else:
                nTrees += 1;
    # выводим ответ:
    minAnsw = nCycles;
    if nTrees > 0:
        minAnsw += 1;
    maxAnsw = dsu.numComponents;
    print(minAnsw, maxAnsw);

File: mirea/cources/beginner/connected-components-3/e.cpp
#include <iostream>
#include <vector>
using namespace std;

// Структура для системы непересекающихся множеств
class DSU {
public:
    vector<int> sz, par;
    int numComponents;

    DSU(int n) : numComponents(n) {
        sz.resize(n, 1);
        par.resize(n);
        for (int i = 0; i < n; i++) {
            par[i] = i;
        }
    }

    int getPar(int u) {
        if (u == par[u]) return u;
        return par[u] = getPar(par[u]); // Сжатие пути
    }

    void connect(int u, int v) {
        u = getPar(u);
        v = getPar(v);
        if (u != v) {
            if (sz[u] < sz[v]) swap(u, v);
            sz[u] += sz[v];
            par[v] = u;
            numComponents--;
        }
    }
};

int main() {
    // Ускорение ввода-вывода
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    
    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        vector<int> a(n);
        for (int i = 0; i < n; i++) {
            cin >> a[i];
            a[i]--;
        }

        DSU dsu(n);
        for (int i = 0; i < n; i++) {
            dsu.connect(i, a[i]);
        }

        // Подсчет минимального ответа
        vector<int> uniqEdges(n, 0);
        vector<bool> isParent(n, false);
        for (int i = 0; i < n; i++) {
            int pi = dsu.getPar(i);
            uniqEdges[pi] = 2 * dsu.sz[pi];
            isParent[pi] = true;
        }

        // Вычитание петель
        for (int u = 0; u < n; u++) {
            if (a[a[u]] == u) {
                uniqEdges[dsu.getPar(u)]--;
            }
        }

        // Подсчет количества циклов и деревьев
        int nCycles = 0, nTrees = 0;
        for (int i = 0; i < n; i++) {
            if (isParent[i]) {
                if (uniqEdges[i] == 2 * dsu.sz[i]) {
                    nCycles++;
                } else {
                    nTrees++;
                }
            }
        }

        int minAnsw = nCycles + (nTrees > 0 ? 1 : 0);
        int maxAnsw = dsu.numComponents;
        cout << minAnsw << " " << maxAnsw << "\n";
    }
    return 0;
}